src/ml/neural_net/combine_base.hpp

```diff
@@ -0,0 +1,253 @@
+ * interface from different threads.
+ */
+template <typename T>
+class Subscriber {
+ public:
+  /** The type of the values that this Subscriber consumes. */
+  using Input = T;
+
+  virtual ~Subscriber() = default;
+
+  /**
+   * The first signal that a Subscriber receives from a Publisher, passing the
+   * Subscription that the Subscriber can use to control the flow of values.
+   *
+   * A Subscriber may only have one Publisher. If it somehow receives more than
+   * one Subscription, it should call Subscription::Cancel() on any instances
+   * received after the first.
+   *
+   * A Subscriber is explictly allowed to demand values synchronously from
+   * within its implementation of this method.
+   */
+  virtual void Receive(std::shared_ptr<Subscription> subscription) = 0;
+
+  /**
+   * Transmits a value from the Publisher to this Subscriber.
+   *
+   * A Subcriber should never receive more calls to this method than the total
+   * Demand it has requested from its publisher. Subscribers should only demand
+   * more elements from within this method via its return value.
+   */
+  virtual Demand Receive(Input element) = 0;
+
+  /**
+   * Signals completion of the stream of values from the Publisher.
+   *
+   * A Subscriber should not receive any further signals of any kind after
+   * receiving a Completion.
+   */
+  virtual void Receive(Completion completion) = 0;
+};
+
+/**
+ * Interface for objects that produce values on demand from its Subscribers.
+ *
+ * Unless otherwise specified by the concrete implementation, external
+ * synchronization must be used to avoid concurrent calls on multiple threads to
+ * a Publisher, including via the Subscriptions that it passes to its
+ * Subscribers.
+ *
+ * Each concrete implementation defines whether it is unicast or multicast:
+ * whether multiple Subscribers observe the same values or not. (An
+ * implementation might only support one Subscriber, by passing an immediate
+ * Completion to each Subscriber after the first.)
+ */
+template <typename T>
+class Publisher : public std::enable_shared_from_this<Publisher<T>> {
+ public:
+  /** The type of values that this Publisher produces. */
+  using Output = T;
+
+  virtual ~Publisher() = default;
+
+  /**
+   * Establishes a connection between this Publisher and the given Subcriber.
+   *
+   * The Publisher must eventually call Subscriber::Receive(Subscription) on the
+   * given Subscriber (and may do so synchronously). The Publisher must then
+   * conform to the protocol established by the Subscription.
+   */
+  virtual void Receive(std::shared_ptr<Subscriber<Output>> subscriber) = 0;
+
+  // Convenienience methods, supporting the chaining together of operations.
+  // Many of these rely on the forward declarations above. Client code should
+  // include combine.hpp to ensure these are defined before they are used.
+
+  void Subscribe(std::shared_ptr<Subscriber<Output>> subscriber) {
```

Comment:
do you mean publisher can subscribe subscriber?

Reply:
Yes, this follows the Swift Combine design from here: https://developer.apple.com/documentation/combine/publisher/3204756-subscribe